From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Mon, 4 Nov 2024 13:59:32 +0800
Subject: [PATCH] Support Virtual Thread

All options can be turned on individually, with more detailed control than Vine.

The Noise part comes from Canvas.

diff --git a/src/main/java/com/mojang/authlib/yggdrasil/YggdrasilServicesKeyInfo.java b/src/main/java/com/mojang/authlib/yggdrasil/YggdrasilServicesKeyInfo.java
index 5dbaf7689363719501d9b41f9537ebd4677e32a6..3eaa893b44e4370bcb04582d08958645a6c57316 100644
--- a/src/main/java/com/mojang/authlib/yggdrasil/YggdrasilServicesKeyInfo.java
+++ b/src/main/java/com/mojang/authlib/yggdrasil/YggdrasilServicesKeyInfo.java
@@ -34,6 +34,7 @@ public class YggdrasilServicesKeyInfo implements ServicesKeyInfo {
     private static final Logger LOGGER = LoggerFactory.getLogger(YggdrasilServicesKeyInfo.class);
 
     private static final ScheduledExecutorService FETCHER_EXECUTOR = Executors.newScheduledThreadPool(1, new ThreadFactoryBuilder()
+        .setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.network.YggdrasilServices.getThreadFactory())
         .setNameFormat("Yggdrasil Key Fetcher")
         .setDaemon(true)
         .build()
diff --git a/src/main/java/io/papermc/paper/pluginremap/PluginRemapper.java b/src/main/java/io/papermc/paper/pluginremap/PluginRemapper.java
index 28857d0c9b53f2068d51b8f09ef40df7a2b97502..b46133e6a37a9547e31a1b2a14946d4504086698 100644
--- a/src/main/java/io/papermc/paper/pluginremap/PluginRemapper.java
+++ b/src/main/java/io/papermc/paper/pluginremap/PluginRemapper.java
@@ -429,6 +429,7 @@ public final class PluginRemapper {
             TimeUnit.SECONDS,
             ScalingThreadPool.createUnboundedQueue(),
             new ThreadFactoryBuilder()
+                .setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.paper.PluginRemapper.getThreadFactory())
                 .setNameFormat("Paper Plugin Remapper Thread - %1$d")
                 .setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandlerWithName(LOGGER))
                 .build(),
diff --git a/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaAsyncScheduler.java b/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaAsyncScheduler.java
index 374abffb9f1ce1a308822aed13038e77fe9ca08b..77be3946a5561c77cf5e2f38868fc800b6afc3e2 100644
--- a/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaAsyncScheduler.java
+++ b/src/main/java/io/papermc/paper/threadedregions/scheduler/FoliaAsyncScheduler.java
@@ -31,7 +31,7 @@ public final class FoliaAsyncScheduler implements AsyncScheduler {
 
             @Override
             public Thread newThread(final Runnable run) {
-                final Thread ret = new Thread(run);
+                final Thread ret = one.tranic.sewlia.config.mod.feature.virtual_thread.folia.AsyncScheduler.getThread(run);
 
                 ret.setName("Folia Async Scheduler Thread #" + this.idGenerator.getAndIncrement());
                 ret.setPriority(Thread.NORM_PRIORITY - 1);
@@ -47,7 +47,7 @@ public final class FoliaAsyncScheduler implements AsyncScheduler {
     private final ScheduledExecutorService timerThread = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
         @Override
         public Thread newThread(final Runnable run) {
-            final Thread ret = new Thread(run);
+            final Thread ret = one.tranic.sewlia.config.mod.feature.virtual_thread.folia.AsyncScheduler.getThread(run);
 
             ret.setName("Folia Async Scheduler Thread Timer");
             ret.setPriority(Thread.NORM_PRIORITY + 1);
diff --git a/src/main/java/io/papermc/paper/util/MCUtil.java b/src/main/java/io/papermc/paper/util/MCUtil.java
index 4dcec640f5870d713bd3b98389a45dbef8a4ea8a..4a9844d825934137b8b9c2f7e17cd51a555539f6 100644
--- a/src/main/java/io/papermc/paper/util/MCUtil.java
+++ b/src/main/java/io/papermc/paper/util/MCUtil.java
@@ -38,6 +38,7 @@ public final class MCUtil {
         }
     };
     public static final ExecutorService ASYNC_EXECUTOR = Executors.newFixedThreadPool(2, new ThreadFactoryBuilder()
+        .setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.paper.AsyncTaskHandler.getThreadFactory())
         .setNameFormat("Paper Async Task Handler Thread - %1$d")
         .setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(MinecraftServer.LOGGER))
         .build()
diff --git a/src/main/java/io/papermc/paper/util/ServerWorkerThreadFactory.java b/src/main/java/io/papermc/paper/util/ServerWorkerThreadFactory.java
new file mode 100644
index 0000000000000000000000000000000000000000..0dab3b428b404f4f844a85c153a337305c4cec9b
--- /dev/null
+++ b/src/main/java/io/papermc/paper/util/ServerWorkerThreadFactory.java
@@ -0,0 +1,30 @@
+package io.papermc.paper.util;
+
+import net.minecraft.Util;
+import one.tranic.sewlia.config.mod.feature.virtual_thread.paper.Worker;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ServerWorkerThreadFactory implements ThreadFactory {
+    private static final AtomicInteger threadId = new AtomicInteger(1);
+    private final String poolName;
+    private final int prioritityModifier;
+
+    public ServerWorkerThreadFactory(String poolName, int prioritityModifier) {
+        this.poolName = poolName;
+        this.prioritityModifier = prioritityModifier;
+    }
+
+    @Override
+    public Thread newThread(@NotNull Runnable r) {
+        final Thread ret = Worker.value ? Thread.ofVirtual().unstarted(r) : new Thread(r);
+        ret.setName("Worker-" + poolName + "-" + threadId.getAndIncrement());
+        ret.setPriority(Thread.NORM_PRIORITY+prioritityModifier); // Deprioritize over main
+        ret.setDaemon(true);
+        ret.setUncaughtExceptionHandler(Util::onThreadException);
+
+        return ret;
+    }
+}
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/Util.java b/src/main/java/net/minecraft/Util.java
index 38ffa1c4c204f73746c4b87127de92b631548eb1..bf15e0b095c88d9b206881a0fe2a361b7d2c7f61 100644
--- a/src/main/java/net/minecraft/Util.java
+++ b/src/main/java/net/minecraft/Util.java
@@ -93,9 +93,9 @@ public class Util {
     private static final int DEFAULT_SAFE_FILE_OPERATION_RETRIES = 10;
     private static final String MAX_THREADS_SYSTEM_PROPERTY = "max.bg.threads";
     private static final TracingExecutor BACKGROUND_EXECUTOR = makeExecutor("Main", -1); // Paper - Perf: add priority
-    private static final TracingExecutor IO_POOL = makeIoExecutor("IO-Worker-", false);
+    private static final TracingExecutor IO_POOL = makeIoExecutor("IO-Worker-", false, one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.io.Worker.value); // Sewlia - VT2.0
     public static final TracingExecutor DIMENSION_DATA_IO_POOL = makeExtraIoExecutor("Dimension-Data-IO-Worker-"); // Paper - Separate dimension data IO pool
-    private static final TracingExecutor DOWNLOAD_POOL = makeIoExecutor("Download-", true);
+    private static final TracingExecutor DOWNLOAD_POOL = makeIoExecutor("Download-", true, one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.network.Download.value); // Sewlia - VT2.0
     // Paper start - don't submit BLOCKING PROFILE LOOKUPS to the world gen thread
     public static final ExecutorService PROFILE_EXECUTOR = Executors.newFixedThreadPool(2, new java.util.concurrent.ThreadFactory() {
 
@@ -103,7 +103,7 @@ public class Util {
 
         @Override
         public Thread newThread(Runnable run) {
-            Thread ret = new Thread(run);
+            Thread ret = one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.network.ProfileLookup.getThread(run); // Sewlia - VT2.0
             ret.setName("Profile Lookup Executor #" + this.count.getAndIncrement());
             ret.setUncaughtExceptionHandler((Thread thread, Throwable throwable) -> {
                 LOGGER.error("Uncaught exception in thread " + thread.getName(), throwable);
@@ -172,7 +172,7 @@ public class Util {
         if (i <= 0) {
             executorService = MoreExecutors.newDirectExecutorService();
         } else {
-            executorService = Executors.newFixedThreadPool(i, target -> new io.papermc.paper.util.ServerWorkerThread(target, name, priorityModifier));
+            executorService = Executors.newFixedThreadPool(i , new io.papermc.paper.util.ServerWorkerThreadFactory(name, priorityModifier)); // Sewlia - VT2.0
         }
         /*      final String string2 = "Worker-" + name + "-" + atomicInteger.getAndIncrement();
                 ForkJoinWorkerThread forkJoinWorkerThread = new ForkJoinWorkerThread(pool) {
@@ -253,10 +253,15 @@ public class Util {
         IO_POOL.shutdownAndAwait(3L, TimeUnit.SECONDS);
     }
 
+    // Sewlia start - VT2.0
     private static TracingExecutor makeIoExecutor(String namePrefix, boolean daemon) {
+        return makeIoExecutor(namePrefix, daemon, false);
+    }
+
+    private static TracingExecutor makeIoExecutor(String namePrefix, boolean daemon, boolean virtual) {
         AtomicInteger atomicInteger = new AtomicInteger(1);
         return new TracingExecutor(Executors.newCachedThreadPool(runnable -> {
-            Thread thread = new Thread(runnable);
+            Thread thread = virtual ? Thread.ofVirtual().unstarted(runnable) : new Thread(runnable);
             String string2 = namePrefix + atomicInteger.getAndIncrement();
             TracyClient.setThreadName(string2, namePrefix.hashCode());
             thread.setName(string2);
@@ -264,13 +269,13 @@ public class Util {
             thread.setUncaughtExceptionHandler(Util::onThreadException);
             return thread;
         }));
-    }
+    } // Sewlia end - VT2.0
 
     // Paper start - Separate dimension data IO pool
     private static TracingExecutor makeExtraIoExecutor(String namePrefix) {
         AtomicInteger atomicInteger = new AtomicInteger(1);
         return new TracingExecutor(Executors.newFixedThreadPool(4, runnable -> {
-            Thread thread = new Thread(runnable);
+            Thread thread = one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.io.DimensionData.value ? Thread.ofVirtual().unstarted(runnable) : new Thread(runnable); // Sewlia - VT2.0
             String string2 = namePrefix + atomicInteger.getAndIncrement();
             TracyClient.setThreadName(string2, namePrefix.hashCode());
             thread.setName(string2);
diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index c54b897b0e14ac974a903729918a45a9eeec236b..6a2d6a3f8026ef13743b761a78f64a97d4621380 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -486,6 +486,7 @@ public class Commands {
         2, 2, 0, java.util.concurrent.TimeUnit.MILLISECONDS,
         new java.util.concurrent.LinkedBlockingQueue<>(),
         new com.google.common.util.concurrent.ThreadFactoryBuilder()
+            .setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.paper.AsyncCommandBuilder.getThreadFactory())
             .setNameFormat("Paper Async Command Builder Thread Pool - %1$d")
             .setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(net.minecraft.server.MinecraftServer.LOGGER))
             .build(),
diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index 30e3d0d7b0ca8cfa30efb60eefa73d94c65c4690..fd0b8d53f37fd1ab250e636a163661dae9ee8414 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -82,13 +82,13 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
         marker.add(Connection.PACKET_MARKER);
     });
     public static final Supplier<NioEventLoopGroup> NETWORK_WORKER_GROUP = Suppliers.memoize(() -> {
-        return new NioEventLoopGroup(0, (new ThreadFactoryBuilder()).setNameFormat("Netty Client IO #%d").setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()); // Paper
+        return new NioEventLoopGroup(0, (new ThreadFactoryBuilder()).setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.netty.Client.getThreadFactory()).setNameFormat("Netty Client IO #%d").setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()); // Paper
     });
     public static final Supplier<EpollEventLoopGroup> NETWORK_EPOLL_WORKER_GROUP = Suppliers.memoize(() -> {
-        return new EpollEventLoopGroup(0, (new ThreadFactoryBuilder()).setNameFormat("Netty Epoll Client IO #%d").setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()); // Paper
+        return new EpollEventLoopGroup(0, (new ThreadFactoryBuilder()).setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.netty.EpollClient.getThreadFactory()).setNameFormat("Netty Epoll Client IO #%d").setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()); // Paper
     });
     public static final Supplier<DefaultEventLoopGroup> LOCAL_WORKER_GROUP = Suppliers.memoize(() -> {
-        return new DefaultEventLoopGroup(0, (new ThreadFactoryBuilder()).setNameFormat("Netty Local Client IO #%d").setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()); // Paper
+        return new DefaultEventLoopGroup(0, (new ThreadFactoryBuilder()).setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.netty.LocalClient.getThreadFactory()).setNameFormat("Netty Local Client IO #%d").setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()); // Paper
     });
     private static final ProtocolInfo<ServerHandshakePacketListener> INITIAL_PROTOCOL = HandshakeProtocols.SERVERBOUND;
     private final PacketFlow receiving;
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 5495f5e2236d241dba223849df40ab0831120f4c..ff52d1381692bcd325be97300d4ed7dc5bdd6d01 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -3133,7 +3133,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     }
 
     public final java.util.concurrent.ExecutorService chatExecutor = java.util.concurrent.Executors.newCachedThreadPool(
-            new com.google.common.util.concurrent.ThreadFactoryBuilder().setDaemon(true).setNameFormat("Async Chat Thread - #%d").setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(net.minecraft.server.MinecraftServer.LOGGER)).build()); // Paper
+            new com.google.common.util.concurrent.ThreadFactoryBuilder().setDaemon(true).setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.AsyncChat.getThreadFactory()).setNameFormat("Async Chat Thread - #%d").setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(net.minecraft.server.MinecraftServer.LOGGER)).build()); // Paper
 
     public final ChatDecorator improvedChatDecorator = new io.papermc.paper.adventure.ImprovedChatDecorator(this); // Paper - adventure
     public ChatDecorator getChatDecorator() {
diff --git a/src/main/java/net/minecraft/server/chase/ChaseClient.java b/src/main/java/net/minecraft/server/chase/ChaseClient.java
index b38c3491911277decaaddc8004ceccf2eeb0e3a4..0fcb359cb728360c0c5bc8e1e1ced812ecd6c200 100644
--- a/src/main/java/net/minecraft/server/chase/ChaseClient.java
+++ b/src/main/java/net/minecraft/server/chase/ChaseClient.java
@@ -51,7 +51,7 @@ public class ChaseClient {
         }
 
         this.wantsToRun = true;
-        this.thread = new Thread(this::run, "chase-client");
+        this.thread = one.tranic.sewlia.config.mod.feature.virtual_thread.chase.Client.getThread(this::run, "chase-client");
         this.thread.setDaemon(true);
         this.thread.start();
     }
diff --git a/src/main/java/net/minecraft/server/chase/ChaseServer.java b/src/main/java/net/minecraft/server/chase/ChaseServer.java
index f3610334289854e949b4ace158554015d85aba54..c6bbf8b033c479c81ea99303e07da8c0dfc0feaf 100644
--- a/src/main/java/net/minecraft/server/chase/ChaseServer.java
+++ b/src/main/java/net/minecraft/server/chase/ChaseServer.java
@@ -44,10 +44,10 @@ public class ChaseServer {
         } else {
             this.wantsToRun = true;
             this.serverSocket = new ServerSocket(this.serverPort, 50, InetAddress.getByName(this.serverBindAddress));
-            Thread thread = new Thread(this::runAcceptor, "chase-server-acceptor");
+            Thread thread = one.tranic.sewlia.config.mod.feature.virtual_thread.chase.ServerAcceptor.getThread(this::runAcceptor, "chase-server-acceptor");
             thread.setDaemon(true);
             thread.start();
-            Thread thread2 = new Thread(this::runSender, "chase-server-sender");
+            Thread thread2 = one.tranic.sewlia.config.mod.feature.virtual_thread.chase.ServerSender.getThread(this::runSender, "chase-server-sender");
             thread2.setDaemon(true);
             thread2.start();
         }
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index a88addb2d75f5f27627ee1a557956d994922b521..5e3043e314e82683f80e748032f9a3a4ec41ec94 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -104,58 +104,15 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     @Override
     public boolean initServer() throws IOException {
-        Thread thread = new Thread("Server console handler") {
-            public void run() {
+        Thread thread = one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.Console.getThread("Server console handler" ,() -> {
                 // CraftBukkit start
                 if (!org.bukkit.craftbukkit.Main.useConsole) {
                     return;
                 }
                 // Paper start - Use TerminalConsoleAppender
                 new com.destroystokyo.paper.console.PaperConsole(DedicatedServer.this).start();
-                /*
-                jline.console.ConsoleReader bufferedreader = DedicatedServer.this.reader;
-
-                // MC-33041, SPIGOT-5538: if System.in is not valid due to javaw, then return
-                try {
-                    System.in.available();
-                } catch (IOException ex) {
-                    return;
-                }
-                // CraftBukkit end
-
-                String s;
-
-                try {
-                    // CraftBukkit start - JLine disabling compatibility
-                    while (!DedicatedServer.this.isStopped() && DedicatedServer.this.isRunning()) {
-                        if (org.bukkit.craftbukkit.Main.useJline) {
-                            s = bufferedreader.readLine(">", null);
-                        } else {
-                            s = bufferedreader.readLine();
-                        }
-
-                        // SPIGOT-5220: Throttle if EOF (ctrl^d) or stdin is /dev/null
-                        if (s == null) {
-                            try {
-                                Thread.sleep(50L);
-                            } catch (InterruptedException ex) {
-                                Thread.currentThread().interrupt();
-                            }
-                            continue;
-                        }
-                        if (s.trim().length() > 0) { // Trim to filter lines which are just spaces
-                            DedicatedServer.this.issueCommand(s, DedicatedServer.this.getServerCommandListener());
-                        }
-                        // CraftBukkit end
-                    }
-                } catch (IOException ioexception) {
-                    DedicatedServer.LOGGER.error("Exception handling console input", ioexception);
-                }
-
-                */
                 // Paper end
-            }
-        };
+        });
 
         // CraftBukkit start - TODO: handle command-line logging arguments
         java.util.logging.Logger global = java.util.logging.Logger.getLogger("");
@@ -344,7 +301,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             }
 
             if (false && this.getMaxTickLength() > 0L) {  // Spigot - disable
-                Thread thread1 = new Thread(new ServerWatchdog(this));
+                Thread thread1 = one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.ServerWatchdog.getThread(new ServerWatchdog(this));
 
                 thread1.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandlerWithName(DedicatedServer.LOGGER));
                 thread1.setName("Server Watchdog");
@@ -480,7 +437,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
                     sender.sendMessage(net.kyori.adventure.text.Component.text("Failed to dump chunk information, see console", net.kyori.adventure.text.format.NamedTextColor.RED));
                 }
             };
-            Thread t = new Thread(run);
+            Thread t = one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.AsyncDebug.getThread(run);
             t.setName("Async debug thread #" + ASYNC_DEBUG_CHUNKS_COUNT.getAndIncrement());
             t.setDaemon(true);
             t.start();
diff --git a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
index 759062d219ff490a3cb19e710c4d18e3e08288e0..a11779343186fad9f3298f3c3d87af3a29757ae3 100644
--- a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
@@ -181,7 +181,7 @@ public class MinecraftServerGui extends JComponent {
         jpanel.add(jscrollpane, "Center");
         jpanel.add(jtextfield, "South");
         jpanel.setBorder(new TitledBorder(new EtchedBorder(), "Log and chat"));
-        this.logAppenderThread = new Thread(() -> {
+        this.logAppenderThread = one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.Gui.getThread(() -> {
             String s;
 
             while ((s = LogQueues.getNextLogEvent("ServerGuiConsole")) != null) {
diff --git a/src/main/java/net/minecraft/server/network/ServerConnectionListener.java b/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
index ae88c6e2635b1608383f8c74813d723f7c6ffaf8..fd090b21fcb759a851359ce564e0424b11800ef4 100644
--- a/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
+++ b/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
@@ -52,10 +52,10 @@ public class ServerConnectionListener {
 
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final Supplier<NioEventLoopGroup> SERVER_EVENT_GROUP = Suppliers.memoize(() -> {
-        return new NioEventLoopGroup(0, (new ThreadFactoryBuilder()).setNameFormat("Netty Server IO #%d").setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()); // Paper
+        return new NioEventLoopGroup(0, (new ThreadFactoryBuilder()).setNameFormat("Netty Server IO #%d").setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.netty.Server.getThreadFactory()).setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()); // Paper
     });
     public static final Supplier<EpollEventLoopGroup> SERVER_EPOLL_EVENT_GROUP = Suppliers.memoize(() -> {
-        return new EpollEventLoopGroup(0, (new ThreadFactoryBuilder()).setNameFormat("Netty Epoll Server IO #%d").setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()); // Paper
+        return new EpollEventLoopGroup(0, (new ThreadFactoryBuilder()).setNameFormat("Netty Epoll Server IO #%d").setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.netty.EpollServer.getThreadFactory()).setDaemon(true).setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(LOGGER)).build()); // Paper
     });
     final MinecraftServer server;
     public volatile boolean running;
diff --git a/src/main/java/net/minecraft/server/rcon/thread/GenericThread.java b/src/main/java/net/minecraft/server/rcon/thread/GenericThread.java
index 3691e97e83e4f4883adc85ec2d8aa789812bd474..6120d2d6f49c971c367d92921fe33788787391e0 100644
--- a/src/main/java/net/minecraft/server/rcon/thread/GenericThread.java
+++ b/src/main/java/net/minecraft/server/rcon/thread/GenericThread.java
@@ -24,7 +24,7 @@ public abstract class GenericThread implements Runnable {
             return true;
         } else {
             this.running = true;
-            this.thread = new Thread(this, this.name + " #" + UNIQUE_THREAD_ID.incrementAndGet());
+            this.thread = one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.network.RconGeneric.getThread(this, this.name + " #" + UNIQUE_THREAD_ID.incrementAndGet());
             this.thread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandlerWithName(LOGGER));
             this.thread.start();
             LOGGER.info("Thread {} started", this.name);
diff --git a/src/main/java/net/minecraft/util/datafix/DataFixers.java b/src/main/java/net/minecraft/util/datafix/DataFixers.java
index f5bdd1dc3b67e8d5a1aabbb3ea37331d0b72d98a..f33af07a35a400b2417f48beef6e02e0b3484414 100644
--- a/src/main/java/net/minecraft/util/datafix/DataFixers.java
+++ b/src/main/java/net/minecraft/util/datafix/DataFixers.java
@@ -142,7 +142,7 @@ public class DataFixers {
         if (requiredTypes.isEmpty()) {
             return CompletableFuture.completedFuture((Object) null);
         } else {
-            ExecutorService executorservice = Executors.newSingleThreadExecutor((new ThreadFactoryBuilder()).setNameFormat("Datafixer Bootstrap").setDaemon(true).setPriority(1).build());
+            ExecutorService executorservice = Executors.newSingleThreadExecutor((new ThreadFactoryBuilder()).setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.Datafixer.getThreadFactory()).setNameFormat("Datafixer Bootstrap").setDaemon(true).setPriority(1).build());
 
             return DataFixers.DATA_FIXER.optimize(requiredTypes, executorservice);
         }
diff --git a/src/main/java/net/minecraft/util/thread/NamedThreadFactory.java b/src/main/java/net/minecraft/util/thread/NamedThreadFactory.java
index 04460b81d6351f63103199a980e463070c233300..11ef161dd06b4f2cbd8e3b51abe7842cb886bbc5 100644
--- a/src/main/java/net/minecraft/util/thread/NamedThreadFactory.java
+++ b/src/main/java/net/minecraft/util/thread/NamedThreadFactory.java
@@ -19,7 +19,7 @@ public class NamedThreadFactory implements ThreadFactory {
 
     @Override
     public Thread newThread(Runnable runnable) {
-        Thread thread = new Thread(this.group, runnable, this.namePrefix + this.threadNumber.getAndIncrement(), 0L);
+        Thread thread = one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.NamedThread.getThread(this.group, runnable, this.namePrefix + this.threadNumber.getAndIncrement(), 0L);
         thread.setUncaughtExceptionHandler((threadx, throwable) -> {
             LOGGER.error("Caught exception in thread {} from {}", threadx, runnable);
             LOGGER.error("", throwable);
diff --git a/src/main/java/net/minecraft/util/worldupdate/WorldUpgrader.java b/src/main/java/net/minecraft/util/worldupdate/WorldUpgrader.java
index 622d0cbe023774d92d212f242b60b96317720835..049cd144434217d4ef1a5de97e300c07bd0781db 100644
--- a/src/main/java/net/minecraft/util/worldupdate/WorldUpgrader.java
+++ b/src/main/java/net/minecraft/util/worldupdate/WorldUpgrader.java
@@ -51,7 +51,7 @@ import org.slf4j.Logger;
 public class WorldUpgrader implements AutoCloseable {
 
     static final Logger LOGGER = LogUtils.getLogger();
-    private static final ThreadFactory THREAD_FACTORY = (new ThreadFactoryBuilder()).setDaemon(true).build();
+    private static final ThreadFactory THREAD_FACTORY = (new ThreadFactoryBuilder()).setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.minecraft.io.WorldUpgrader.getThreadFactory()).setDaemon(true).build();
     private static final String NEW_DIRECTORY_PREFIX = "new_";
     static final Component STATUS_UPGRADING_POI = Component.translatable("optimizeWorld.stage.upgrading.poi");
     static final Component STATUS_FINISHED_POI = Component.translatable("optimizeWorld.stage.finished.poi");
diff --git a/src/main/java/org/bukkit/craftbukkit/bootstrap/Main.java b/src/main/java/org/bukkit/craftbukkit/bootstrap/Main.java
index 8a4f95049c63afb28bef6719c77b7a7092e75aae..3a7ba39c0bf59443392eacae547c7a58ed14f653 100644
--- a/src/main/java/org/bukkit/craftbukkit/bootstrap/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/bootstrap/Main.java
@@ -53,7 +53,7 @@ public class Main {
             URLClassLoader classLoader = new URLClassLoader(extractedUrls.toArray(new URL[0]));
 
             System.out.println("Starting server");
-            Thread runThread = new Thread(() -> {
+            Thread runThread = one.tranic.sewlia.config.mod.feature.virtual_thread.bukkit.ServerMain.getThread(() -> {
                 try {
                     Class<?> mainClass = Class.forName(mainClassName, true, classLoader);
                     MethodHandle mainHandle = MethodHandles.lookup().findStatic(mainClass, "main", MethodType.methodType(void.class, String[].class)).asFixedArity();
diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftAsyncScheduler.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftAsyncScheduler.java
index 0ca279fb71d39c81b1f608e0ee9ba3e498d55fa3..7e8078b3a2f3e145fbff0711c2db81618bb6bebe 100644
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftAsyncScheduler.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftAsyncScheduler.java
@@ -40,9 +40,9 @@ public class CraftAsyncScheduler extends CraftScheduler {
 
     private final ThreadPoolExecutor executor = new ThreadPoolExecutor(
             4, Integer.MAX_VALUE,30L, TimeUnit.SECONDS, new SynchronousQueue<>(),
-            new ThreadFactoryBuilder().setNameFormat("Craft Scheduler Thread - %1$d").build());
+            new ThreadFactoryBuilder().setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.bukkit.CraftScheduler.getThreadFactory()).setNameFormat("Craft Scheduler Thread - %1$d").build());
     private final Executor management = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
-            .setNameFormat("Craft Async Scheduler Management Thread").build());
+            .setThreadFactory(one.tranic.sewlia.config.mod.feature.virtual_thread.bukkit.AsyncScheduler.getThreadFactory()).setNameFormat("Craft Async Scheduler Management Thread").build());
     private final List<CraftTask> temp = new ArrayList<>();
 
     CraftAsyncScheduler() {
